### PR TITLE
Add polar coordinate `sf::Vector2<T>` constructor

### PR DESCRIPTION
**Pull request by ChrisThrasher**
## Description

This is the corollary to [`sf::Vector2<T>::angle()`](https://github.com/SFML/SFML/blob/master/include/SFML/System/Vector2.hpp#L121) which produces an angle from a vector. To avoid circular dependencies and to ensure we can construct vectors of all types from angles, I made this a new `sf::Vector2<T>` constructor overload instead of adding a new member function to `sf::Angle`. 

The motivation for this was simplifying how we currently create unit vectors from angles which looks like [`sf::Vector2f::UnitX.rotatedBy(rotation)`](https://github.com/ChrisThrasher/boids/blob/sfml-v3/src/boid.cpp#L23). I wouldn't expect many users to realize this is possible since at first glance it's not obvious what I'm trying to do here.

As an aside, this PR also removes the degree sign from Vector2.hpp which is an extended ASCII character. SFML appears to avoid extended ASCII characters in source files so I wanted to bring this file in line with the rest of the library.

## Tasks

* [x] Tested on Linux
* [x] Tested on Windows
* [x] Tested on macOS
* [x] Tested on iOS
* [x] Tested on Android

### CODE REVIEW

**Comment by Bromeon** [comment]
Nice idea! I think the constructor should take angle **and length** though.
Basically the polar-coordinate equivalent `(r, phi)` to cartesian `(x, y)`.

Also, we should disallow construction from angles for non-float vector types.

**Comment by ChrisThrasher** [comment]
> Nice idea! I think the constructor should take angle **and length** though. Basically the polar-coordinate equivalent `(r, phi)` to cartesian `(x, y)`.

Would you rather the length be an optional parameter or keep two constructors, one for unit vectors and one for vectors with a specified length?

> Also, we should disallow construction from angles for non-float vector types.

Good point. I'll fix that right now.

**Comment by Bromeon** [comment]
> Would you rather the length be an optional parameter or keep two constructors, one for unit vectors and one for vectors with a specified length?

Personally I would not treat the `(1, phi)` construction as a special case, as we don't do that for any value of `x`, `y` or `phi` either. 

A general `(r, phi)` constructor should be a good trade-off regarding flexibility and API surface.

File: include/SFML/System/Vector2.hpp
Comment (Bromeon) [inline include/SFML/System/Vector2.hpp:83]: I'd probably pass `Angle` by value and name it `phi` like here:
https://github.com/SFML/SFML/blob/c7327353b3eaa1ac62c13195a1424033121fb5de/include/SFML/System/Vector2.hpp#L140

(also applies once the constructor signature changes)

Comment (ChrisThrasher) [reply include/SFML/System/Vector2.hpp:83]: Weird, the implementation uses the name `angle`. I'll use `phi` if you want though.

https://github.com/SFML/SFML/blob/3ae85854e4674debe3b30ae2f6cf9d9926738789/include/SFML/System/Vector2.inl#L110

Comment (Bromeon) [reply include/SFML/System/Vector2.hpp:83]: Oh, that's an oversight, feel free to fix it. Ultimately it's not _that_ important 🙂

Comment (ChrisThrasher) [reply include/SFML/System/Vector2.hpp:83]: Might be worth some library-wide rename alongside changing many `const Angle&` params to `Angle`. For a later PR but I'll make a note to check for consistent `sf::Angle` use.

**Comment by ChrisThrasher** [comment]
@Bromeon I changed that constructor to use a length/angle pair and added some new tests cases to make sure everything working as intended. I agree this is a better plan than two constructors and/or something involving optional parameters.

File: include/SFML/System/Vector2.hpp
Comment (Bromeon) [reply include/SFML/System/Vector2.hpp]: I was mostly thinking about calling an angle `phi` in the context of polar coordinates (i.e. a vector) -- i.e. in cases where there is another coordinate `r` (radius). Not sure if that necessarily applies to the rest of the code base; so `angle` might still be fine elsewhere.

File: examples/tennis/Tennis.cpp
Comment (Bromeon) [inline examples/tennis/Tennis.cpp:203]: Nice one! This also shows the strength of a type-safe `sf::Angle` type, otherwise we would have needed something like `sf::Vector2f::fromPolar(r, phi)`.

Are people generally OK with this being a constructor, i.e. allowing short syntax?

File: test/System/Vector2.cpp
Comment (Bromeon) [inline test/System/Vector2.cpp:44]: Could we maybe add some arbitrary values for `phi` here (not just half-circles), to see if it's still the zero vector?
Possibly we need `ApproxVec` for rounding errors.

File: include/SFML/System/Vector2.hpp
Comment (Bromeon) [inline include/SFML/System/Vector2.hpp:84]: Do we want to allow negative values of `length`? (as it's currently handled in tests)

* If yes, we should probably rename the parameter, since it's no longer a length and it won't match the getter `length()`. Maybe `r`.
* Also, in that case we should document that it is allowed, and maybe even that `Vector2(r, phi)` is equivalent to `Vector2(-r, phi + 180°)`.
* Should we mention something about `Vector2(r, phi)` being equivalent to `Vector2(r, phi + n * 360°)`? 

Those things may sound obvious, but I think the important message is that the polar angle constructor is lossy, and it may not be able to retrieve the values later as they were provided in the constructor.

Comment (Bromeon) [reply include/SFML/System/Vector2.hpp:84]: Furthermore, maybe we should change the `\brief` part to:
```cpp
/// \brief Construct the vector from polar coordinates
```
while for the `(x, y)` constructor, we could say:
```cpp
/// \brief Construct the vector from cartesian coordinates
```

File: test/System/Vector2.cpp
Comment (Bromeon) [inline test/System/Vector2.cpp:80]: Is there a particular reason why you use "nearly Pi"? It doesn't seem like you make use of any of its properties, so I wonder if another arbitrary magic number might be less suggestive.

File: test/System/Vector2.cpp
Comment (ChrisThrasher) [reply test/System/Vector2.cpp]: Fixed

File: test/System/Vector2.cpp
Comment (ChrisThrasher) [reply test/System/Vector2.cpp]: Haha yeah that was a bad idea. I'll change that.

File: include/SFML/System/Vector2.hpp
Comment (ChrisThrasher) [reply include/SFML/System/Vector2.hpp]: > Do we want to allow negative values of length? (as it's currently handled in tests)

In the Tennis example I modified, imagine if you wanted the ball to go backwards along a certain trajectory. In that case it's really helpful to use a negative radius/length instead of having extra logic to flip the angle. Besides, the math is perfectly well defined for all values of `r` and `phi`.

> If yes, we should probably rename the parameter, since it's no longer a length and it won't match the getter length(). Maybe r.

No problem. I'll adopt that r/phi terminology. I think it looks good and I like how it helps avoid confusion with `length()` and `angle()` which as you pointed out won't necessarily return the same values that were provided to this constructor. 

> Should we mention something about Vector2(r, phi) being equivalent to Vector2(r, phi + n * 360°)?

No harm in a little extra documentation :)

**Comment by ChrisThrasher** [comment]
Rebasing on master and fixing the new compiler warnings added by #2070 would require making changes that #2089 renders unnecessary so I'd prefer to block this PR until a decision can be reached on #2089.

**Comment by ChrisThrasher** [comment]
@Bromeon Anything more you want to discuss about this?

**Review by Bromeon — APPROVED** [review]
Looks good to me!

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
Rebased after #2055 was merged just to be sure that there we no new conflicts.

**Comment by eXpl0it3r** [comment]
Thanks for this addition. Lets see how it will be received/utilized. 🙂

**Pull request merged**
